Added a rule for having commands with arguments in sudoers - ANSSI R63

matejak opened the pull request:
The rule assures that commands in user specs are always followed by at least one argument.

The regular expression is already monstrous, so although the rule is somehow remediable in theory (one could add "" as an argument to commands that don't have one), there is no remediation, as having one that would work reliably would be very difficult.

The rule has some limitations - it doesn't handle commas in arguments well - `command one\,two` would be the correct way of specifying that only the argument `one,two` is allowed to be specified, and the current regex treats that escaped comma as a commands delimiter.

yuumasato commented on linux_os/guide/system/software/sudo/sudoers_no_command_args/tests/simple.fail.sh:
```diff
@@ -0,0 +1,5 @@
+# packages = sudo
+# remediation = none
+
+echo 'jen !fred		ALL, SERVERS = /bin/sh' > /etc/sudoers
```
And `visudo` complains about `jen !fred ALL, servers =`. The User_List is comma separated.
So it should be `jen,!fred ALL,Servers =`.
And `jen, !fred ALL, Servers =` was accepted too.

yuumasato reviewed (changes requested):
The check seems to working as expected, great work on the regex.
I've uncovered a minor issue on the User_List in the test scenarios that don't seem require any change, but would be good to double check.

> The regular expression is already monstrous, so although the rule is somehow remediable in theory (one could add "" as an argument to commands that don't have one), there is no remediation, as having one that would work reliably would be very difficult.

How about adding the note above as a warning? Similar to done in https://github.com/ComplianceAsCode/content/blob/5573ab62e16808bdb1f879422d2cca91c54e65a1/linux_os/guide/system/software/sudo/sudoers_no_command_negation/rule.yml#L41-L43


> The rule has some limitations - it doesn't handle commas in arguments well - `command one\,two` would be the correct way of specifying that only the argument `one,two` is allowed to be specified, and the current regex treats that escaped comma as a commands delimiter.

What happens when there is a command argument with comma? From what I tested it becomes a finding, i.e.: a false positive, which less of a problem than a false negative.
How about adding the limitation of comma handling to the rule as warning? So that people are less surprised when something valid comes up as a finding?

matejak commented:
Thanks, I think that I have addressed your concerns in a separate commit:

- Added warning about remediation and limitations to the rule, added a false positive test.
- Improved the check not to be confused by trailing whitespace, added a test.
- Fixed tests, so they correspond to real sudoers syntax.

yuumasato commented on linux_os/guide/system/software/sudo/sudoers_no_command_args/oval/shared.xml:
```diff
@@ -0,0 +1,25 @@
+           - ',' is a command delimiter, while
+         The last capturing group holds the offending command without args.
+    -->
+    <ind:pattern operation="pattern match">^(?:\s*[^#=]+)=(?:\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:[ \t]+[^,\s]+)+,)*(\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:,|[ \t]$))</ind:pattern>
```
```suggestion
    <ind:pattern operation="pattern match">^(?:\s*[^#=]+)=(?:\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:[ \t]+[^,\s]+)+[ \t]*,)*(\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+[ \t]*(?:,|$))</ind:pattern>
```

There are two changes in the suggestion above:
- the first should allow matching any number of spaces after arguments of correct command calls
  sample test scenario: `nobody ALL=/bin/ls , (!bob,alice) /bin/dog, /bin/cat arg`
- the second allows matching the last command call without arguments nor spaces
  See comment in `simple.fail.sh`.

yuumasato commented on linux_os/guide/system/software/sudo/sudoers_no_command_args/rule.yml:
```diff
@@ -0,0 +1,46 @@
+
+ocil: |-
+    To determine if arguments that commands can be executed with are restricted, run the following command:
+    <pre>$ sudo grep -PR '^(?:\s*[^#=]+)=(?:\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:[ \t]+[^,\s]+)+,)*(\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:,|$))' /etc/sudoers /etc/sudoers.d/</pre>
```
```suggestion
    <pre>$ sudo grep -PR '^(?:\s*[^#=]+)=(?:\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+(?:[ \t]+[^,\s]+)+[ \t]*,)*(\s*(?:\([^\)]+\))?\s*(?!\s*\()[^,\s]+[ \t]*(?:,|$))' /etc/sudoers /etc/sudoers.d/</pre>
```

matejak replied:
I agree with this suggestion, thanks!

yuumasato commented on linux_os/guide/system/software/sudo/sudoers_no_command_args/tests/simple.fail.sh:
```diff
@@ -0,0 +1,5 @@
+# packages = sudo
+# remediation = none
+
+echo 'jen,!fred		ALL,SERVERS = /bin/sh ' > /etc/sudoers
```
The original test scenario `jen,!fred		ALL,SERVERS = /bin/sh` started to fail.

yuumasato commented on controls/anssi.yml:
```diff
@@ -845,7 +845,8 @@ controls:
     title: Explicit arguments in sudo specifications
-    # rules: TBD
+    rules:
+    - sudoers_no_command_args
```
How about changing rule id to `sudoers_explicit_command_args`, per rule title?

matejak replied:
I agree, that would be a more descriptive ID.

yuumasato reviewed (changes requested):
The fix introduced a new issue, see `simple.fail.sh`, I made a suggestion for that.
And also added suggestion to handle spaces after args in compliant commands.

yuumasato reviewed (approved):
Thank you for the rule, great work! :)
